File: web/src/lib/api/workspace.ts
import { invoke } from "./adapter";

export interface WorkspaceFileInfo {
  name: string;
  exists: boolean;
  sizeBytes: number;
  modifiedAt?: number;
  etag?: string;
}

export interface WorkspaceFileContent {
  name: string;
  content: string;
  sizeBytes: number;
  modifiedAt: number;
  etag: string;
}

export interface WorkspaceWriteOutcome extends WorkspaceFileContent {
  backupId?: string;
}

export interface WorkspaceBackupInfo {
  id: string;
  sizeBytes: number;
  createdAt: number;
}

export interface DailyMemoryInfo {
  date: string;
  sizeBytes: number;
  modifiedAt: number;
  etag: string;
  preview: string;
}

export interface DailyMemorySearchResult {
  date: string;
  sizeBytes: number;
  modifiedAt: number;
  etag: string;
  snippet: string;
  matchCount: number;
}

export interface DailyMemoryDeleteOutcome {
  date: string;
  deleted: boolean;
  backupId?: string;
}

export const workspaceApi = {
  listFiles(): Promise<WorkspaceFileInfo[]> {
    return invoke("list_workspace_files");
  },
  readFile(name: string): Promise<WorkspaceFileContent> {
    return invoke("read_workspace_file", { filename: name });
  },
  writeFile(
    name: string,
    content: string,
    expectedEtag?: string,
  ): Promise<WorkspaceWriteOutcome> {
    return invoke("write_workspace_file", {
      filename: name,
      content,
      expectedEtag: expectedEtag ?? null,
    });
  },
  listBackups(name: string): Promise<WorkspaceBackupInfo[]> {
    return invoke("list_workspace_backups", { filename: name });
  },
  restoreBackup(
    name: string,
    backupId: string,
    expectedEtag?: string,
  ): Promise<WorkspaceWriteOutcome> {
    return invoke("restore_workspace_backup", {
      filename: name,
      backupId,
      expectedEtag: expectedEtag ?? null,
    });
  },
  listDailyMemory(): Promise<DailyMemoryInfo[]> {
    return invoke("list_daily_memory_files");
  },
  readDailyMemory(date: string): Promise<WorkspaceFileContent> {
    return invoke("read_daily_memory_file", { date });
  },
  writeDailyMemory(
    date: string,
    content: string,
    expectedEtag?: string,
  ): Promise<WorkspaceWriteOutcome> {
    return invoke("write_daily_memory_file", {
      date,
      content,
      expectedEtag: expectedEtag ?? null,
    });
  },
  searchDailyMemory(query: string): Promise<DailyMemorySearchResult[]> {
    return invoke("search_daily_memory_files", { query });
  },
  deleteDailyMemory(
    date: string,
    expectedEtag: string,
  ): Promise<DailyMemoryDeleteOutcome> {
    return invoke("delete_daily_memory_file", { date, expectedEtag });
  },
};

File: web/src-tauri/src/commands/workspace.rs
#![allow(non_snake_case)]

use crate::workspace::{
    self, DailyMemoryDeleteOutcome, DailyMemoryInfo, DailyMemorySearchResult, WorkspaceBackupInfo,
    WorkspaceFileContent, WorkspaceFileInfo, WorkspaceWriteOutcome,
};

#[tauri::command]
pub async fn list_workspace_files() -> Result<Vec<WorkspaceFileInfo>, String> {
    tauri::async_runtime::spawn_blocking(workspace::list_files)
        .await
        .map_err(|e| format!("Failed to list workspace files: {e}"))?
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn read_workspace_file(filename: String) -> Result<WorkspaceFileContent, String> {
    tauri::async_runtime::spawn_blocking(move || workspace::read_file(&filename))
        .await
        .map_err(|e| format!("Failed to read workspace file: {e}"))?
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn write_workspace_file(
    filename: String,
    content: String,
    expectedEtag: Option<String>,
) -> Result<WorkspaceWriteOutcome, String> {
    tauri::async_runtime::spawn_blocking(move || {
        workspace::write_file(&filename, &content, expectedEtag.as_deref())
    })
    .await
    .map_err(|e| format!("Failed to write workspace file: {e}"))?
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_workspace_backups(filename: String) -> Result<Vec<WorkspaceBackupInfo>, String> {
    tauri::async_runtime::spawn_blocking(move || workspace::list_backups(&filename))
        .await
        .map_err(|e| format!("Failed to list workspace backups: {e}"))?
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn restore_workspace_backup(
    filename: String,
    backupId: String,
    expectedEtag: Option<String>,
) -> Result<WorkspaceWriteOutcome, String> {
    tauri::async_runtime::spawn_blocking(move || {
        workspace::restore_backup(&filename, &backupId, expectedEtag.as_deref())
    })
    .await
    .map_err(|e| format!("Failed to restore workspace backup: {e}"))?
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_daily_memory_files() -> Result<Vec<DailyMemoryInfo>, String> {
    tauri::async_runtime::spawn_blocking(workspace::list_daily_memory)
        .await
        .map_err(|e| format!("Failed to list daily memory files: {e}"))?
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn read_daily_memory_file(date: String) -> Result<WorkspaceFileContent, String> {
    tauri::async_runtime::spawn_blocking(move || workspace::read_daily_memory(&date))
        .await
        .map_err(|e| format!("Failed to read daily memory file: {e}"))?
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn write_daily_memory_file(
    date: String,
    content: String,
    expectedEtag: Option<String>,
) -> Result<WorkspaceWriteOutcome, String> {
    tauri::async_runtime::spawn_blocking(move || {
        workspace::write_daily_memory(&date, &content, expectedEtag.as_deref())
    })
    .await
    .map_err(|e| format!("Failed to write daily memory file: {e}"))?
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn search_daily_memory_files(
    query: String,
) -> Result<Vec<DailyMemorySearchResult>, String> {
    tauri::async_runtime::spawn_blocking(move || workspace::search_daily_memory(&query))
        .await
        .map_err(|e| format!("Failed to search daily memory files: {e}"))?
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn delete_daily_memory_file(
    date: String,
    expectedEtag: Option<String>,
) -> Result<DailyMemoryDeleteOutcome, String> {
    tauri::async_runtime::spawn_blocking(move || {
        workspace::delete_daily_memory(&date, expectedEtag.as_deref())
    })
    .await
    .map_err(|e| format!("Failed to delete daily memory file: {e}"))?
    .map_err(|e| e.to_string())
}
